charts/CSVGenerator.tsx

```diff
@@ -1,91 +1,200 @@
-                        const latestYearValue = dim.yearAndValueOfLatestValueforEntity(
-                            entity
-                        )
-                        if (latestYearValue) {
-                            row.push(
-                                dim.formatYear(first(latestYearValue) as number)
-                            )
-                            value = last(latestYearValue)
-                        } else row.push("")
-                    } else {
-                        value = dim.valueByEntityAndYear.get(entity)?.get(year)
-                    }
-
-                    if (value != null) {
-                        row.push(value)
-                        rowHasSomeValue = true
-                    } else row.push("")
-                })
-
-                // Only add rows which actually have some data in them
-                if (rowHasSomeValue) rows.push(row.map(csvEscape).join(","))
+
+        return {
+            indexingYears: this.yearIsDayVar!.yearsUniq,
+            titleRow: titleRow
+        }
+    }
+
+    @computed get yearBasedCSVParams(): CSVParams {
+        const titleRow = this.baseTitleRow
+        titleRow.push("Year")
+
+        this.csvDimensions.map(dim =>
+            titleRow.push(csvEscape(dim.fullNameWithUnit))
+        )
+
+        return {
+            indexingYears: uniq(
+                flatten(this.csvDimensions.map(d => d.yearsUniq))
+            ),
+            titleRow: titleRow
+        }
+    }
+
+    @computed get csvParams(): CSVParams {
+        return this.dayIndexedCSV
+            ? this.dayBasedCSVParams
+            : this.yearBasedCSVParams
+    }
+
+    entityCode(entity: string) {
+        return this.chart.entityMetaByKey[entity].code ?? ""
+    }
+
+    private formattedYear(year: number) {
+        return this.chart.formatYearFunction(year)
+    }
+
+    private valueForDimensionEntityYear(
+        dim: ChartDimensionWithOwidVariable,
+        entity: string,
+        year: number
+    ) {
+        return dim.valueByEntityAndYear.get(entity)?.get(year)
+    }
+
+    private yearAndValueForSingleYearDimension(
+        dim: ChartDimensionWithOwidVariable,
+        entity: string
+    ): [string, string | number] | null {
+        let latestYear = undefined
+        let latestValue = undefined
+
+        const latestYearValue = dim.yearAndValueOfLatestValueforEntity(entity)
+
+        if (latestYearValue) {
+            latestYear = dim.formatYear(first(latestYearValue) as number)
+            latestValue = last(latestYearValue)
+        }
+
+        if (latestYear !== undefined && latestValue !== undefined) {
+            return [latestYear, latestValue]
+        } else return null
+    }
+
+    private dimensionsValues(entity: string, year: number) {
+        const values: (string | number)[] = []
+        let rowHasSomeValue = false
+
+        this.allValueDimensions.map(dim => {
+            const value = this.valueForDimensionEntityYear(dim, entity, year)
```

Comment:
nit: why the extra method and not just `dim.valueByEntityAndYear.get(entity)?.get(year)`

Reply:
Thought it might help with readability.

Reply:
Will push as is right now to enable review of other PR, but curious to hear your thoughts via Slack.